In [1]:
import numpy as np
import math as math
import matplotlib.pyplot as plt

In [2]:
kb=1.38e-23#Boltzmanns constant, in the unit of J/K
hp=6.626e-34#Planck constant, in the unit of J s
vc=3.0e8#The speed of light, in the unit of m/s
lam1=6.0e-7#lower desired wavelength, in the unit of m
lam2=7.5e-7#higher desired wavelength,in the unit of m

In [3]:
#The peak locates around 5600K according the results in part (a), and the interval is 100K in part a
Tlow=5500
Thigh=5700
tem=np.arange(Tlow,Thigh+1,1)#Temperature in the unit of K
#Temperature array
num_t=len(tem)
print(num_t)

201


In [4]:
b=hp*vc/(lam1*kb*tem)#upper limitation of integral
a=hp*vc/(lam2*kb*tem)#lower limitation of integral
N=100
h=(b-a)/N

In [5]:
def x(lam,temp):
    x=hp*vc/(lam*kb*temp)
    return x

In [6]:
def fx(x):
    fx=(x**3)/((np.exp(x))-1)
    return fx

In [7]:
R=[]
tempR=[]
#create the 2D array
Rt=np.zeros([N+1,N+1],float)
record=0

In [8]:
for t in range(num_t):
    R.append(record)
    tempR.append(Rt)
    
    temp=tem[t]
    at=a[t]
    bt=b[t]
    fa=fx(at)
    fb=fx(bt)
    
    s=0.5*(fa+fb)
 
    Rt[0,0]=s
    for i in range(1,N+1):
        ht=(bt-at)/i
        for k in range(1,i):
            s += fx(at+(k*ht))
        Rt[i,0]=ht*s 
        for m in range(0,i):
            Rt[i,m+1]=Rt[i,m]+(1/((4**(m+1))-1))*(Rt[i,m]-Rt[i-1,m])
    record=Rt[N,N]*15.0/(np.pi**4)

In [9]:
print(Rt)

[[  1.24353453   0.           0.         ...,   0.           0.           0.        ]
 [  1.04750315   0.98215936   0.         ...,   0.           0.           0.        ]
 [  1.05374759   1.05582906   1.06074038 ...,   0.           0.           0.        ]
 ..., 
 [ 51.22712109  51.40305992  51.43824774 ...,  51.44938347   0.           0.        ]
 [ 51.75493975  51.9308793   51.96606726 ...,  51.97720303  51.97720303
    0.        ]
 [ 52.2827605   52.45870075  52.49388885 ...,  52.50502466  52.50502466
   52.50502466]]


In [10]:
#print(R)
print(len(R))


201


In [61]:
#use the formular 6.121 in textbook
gamma=-10#it is negative because we are looking for a maximum value
step=10000
Ra=np.asarray(R)
x1=5500
x2=5700

xpoints=np.arange(len(R))
#ypoints=Ra

for i in range(step):
    if i == 0:
        xpoints[i]=x1
    if i == 1:
        xpoints[i]=x2
    if i == 2:
        xpoints[i]=int(xpoints[i-1]+(gamma*((Ra[i-1]-Ra[i-2])/(xpoints[i-1]-xpoints[i-2]))))

    if i > 2 and i < len(R):
        if xpoints[i-1]-xpoints[i-2] == 0:
            xpoints[i] = xpoints[i-1]
        else:
            xpoints[i]=int(xpoints[i-1]+(gamma*((Ra[i-1]-Ra[i-2])/(xpoints[i-1]-xpoints[i-2]))))
        #xpoints[i]=x_temp//1


In [62]:
print(xpoints)

[5500 5700 5699 5698 5697 5696 5695 5694 5693 5692 5691 5690 5689 5688 5687
 5686 5685 5684 5683 5682 5681 5680 5679 5678 5677 5676 5675 5674 5673 5672
 5671 5670 5669 5668 5667 5666 5665 5664 5663 5662 5661 5660 5659 5658 5657
 5656 5655 5654 5653 5652 5651 5650 5649 5648 5647 5646 5645 5644 5643 5642
 5641 5640 5639 5638 5637 5636 5635 5634 5633 5632 5631 5630 5629 5628 5627
 5626 5625 5624 5623 5622 5621 5620 5619 5618 5617 5616 5615 5614 5613 5612
 5611 5610 5609 5608 5607 5606 5605 5604 5603 5602 5601 5600 5599 5598 5597
 5596 5595 5594 5593 5592 5591 5590 5589 5588 5587 5586 5585 5584 5583 5582
 5581 5580 5579 5578 5577 5576 5575 5574 5573 5572 5571 5570 5569 5568 5567
 5566 5565 5564 5563 5562 5561 5560 5559 5558 5557 5556 5555 5554 5553 5552
 5551 5550 5549 5548 5547 5546 5545 5544 5543 5542 5541 5540 5539 5538 5537
 5536 5535 5534 5533 5532 5531 5530 5529 5528 5527 5526 5525 5524 5523 5522
 5521 5520 5519 5518 5517 5516 5515 5514 5513 5512 5511 5510 5509 5508 5507
 5506 5505 5

In [65]:
print(xpoints[len(xpoints)-1])
print(np.amax(R))
print(tem[5501-5500])
print(R[5501-5500])

5501
8.10539892903
5501
8.10539892903


So,the temperature of maximum efficiency of the light bulb is 5501 K. The maximum efficiency is 8.11
But it is not practical to run a tungsten-filament light bult at this temperature, 
because the melting point of tungsten is about 3700 K.